### ISSUE TITLE
[naming-convention] Static Computed Properties do not enforce the rule

### ISSUE THREAD

**Issue by berickson1** [typescript-eslint/typescript-eslint#3445]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    '@typescript-eslint/naming-convention': [
    'error',
    // Initial default: Strict camel case everywhere -- no getMyURLs() calls -- must be getMyUrls().
    {
      selector: 'default',
      format: ['strictCamelCase'],
    },
  ],
  }
}
```

```TS
interface foo {
  camelCase: string; // Passes Lint
  ['Not_CamelCase']: string; // Should Fail Lint
}
```

**Expected Result**

Static computed properties must satisfy the lint rules. Since this is a string literal inside a computed type, I'd expect it to be checked. (I understand that actual runtime computed types can't work)

**Actual Result**
No Lint Error on line 3.

**Additional Info**
This gives developers a way to "cheat" and bypass the naming-convention rules on objects.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.25.0` |
| `@typescript-eslint/parser`        | `4.25.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `14.15.4` |

**Comment by bradzacher** [comment]
There's a lint rule to enforce people don't use useless computed keys:
[no-useless-computed-key](https://eslint.org/docs/rules/no-useless-computed-key)

It doesn't have an extension rule, so it doesn't support types/enums, but there's no reason that couldn't be added as it seems like a decent value prop.

---

> This gives developers a way to "cheat" and bypass the naming-convention rules on objects.

I'd be very worried for your codebase if an engineer is purposely (and thus maliciously) circumventing lint rules like this.

**Comment by berickson1** [comment]
I did find that rule and will apply it to our codebase (removing the offending code in the process)
We have a process to review any lint exceptions, so occasionally people find shortcuts to bypass rules unfortunately.

Ultimately there's a hole in the lint rule, even if unlikely to abuse, so wanted to see if this was something you care about.

**Comment by bradzacher** [comment]
Computed names are difficult to handle and they quickly devolve into unhandleable things. Whilst a simple case of just literals is pretty easy to check, nothing behind that is really handleable without complex logic.

For example
```ts
const myProperty = 'My_Property';
class Foo {
  [`My_${'property'}`] = 1
  ['My_Property' + ''] = 2
  [myProperty] = 3;
  [String('My_Property')] = 4;
}
```

Are all clearly bad and a human can see that they are - but they're all ***really*** hard for static analysis to pull it apart and understand.

We could certainly handle the pure literal cases - but I do question if the rule should be in the business of doing that deep validation when it's clearly "stupid" code that shouldn't exist in the codebase to begin with.

---

> so occasionally people find shortcuts to bypass rules unfortunately

The thing I'm surprised about (and would be concerned about engineering culture-wise) is that the code was intentionally written to bypass the lint rule, and then the reviewer also just ignored the bad code.

**Comment by berickson1** [comment]
The examples of things that are hard to parse are almost all blatantly trying to do something....undesirable & should be highlighted by a reviewer. A pure literal as a computed property can be missed by a reviewer who is more concerned about more complex code elsewhere in a PR or is unfamiliar with some of the rules in the codebase. In part, lint rules are there to provide guidelines to help steer developers in a direction or away from a pattern. Most of the time people do the "correct thing", but sometimes abuse (intentional or unintentional) happens occasionally.

**Comment by jpike88** [comment]
Would this cover the following case:

```
export enum CaseReason {

// ban this line
	'application-reviewed-rejected' = 'application-reviewed-rejected',

// allow this line
	ApplicationReviewedRejected = 'application-reviewed-rejected',

}

```

The goal is to allow TS to kick in, as typos referencing that first key aren't considered to be wrong.

**Comment by Josh-Cena** [comment]
To clarify the scope, should we fix this by making a `no-useless-computer-key` extension, or by actually checking computed keys with literal values?

**Comment by bradzacher** [comment]
[Currently the rule checks non-computed literal keys for all "members"](https://typescript-eslint.io/play/#ts=4.9.3&jsx=true&sourceType=module&showAST=es&code=MYewdgzgLgBAHjAvDA3gKBjA2gcgIIE4C6AXDAIwA0GMOAQgzmVWgL4DcaaApmAK4BbGADEQIVDVwE8xJBWqZ6jOSw5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9GADa4bDkTRog6JHkQAuvIC+IHUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA), and it provides an escape hatch of the `requiresQuotes` modifier to allow people to ignore names that must be quoted.

The scope of this new check is that we should specifically handle these cases in the rule:
```ts
const foo = {
  ['name']: 1,
  ["name"]: 2,
  [`name`]: 3,
};
// and all variations eg class property, enum member, object type property, ...
```

But we should not go further that that and attempt to analyse an expression to determine a resulting name.
For example we will continue to ignore these cases:
```
const name = 'name';
const foo = {
  ['name' + '']: 1,
  [String('name')]: 2,
  [name]: 3,
  [`${'name'}`]: 4,
  // etc
};
```

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
